hudi-client/hudi-client-common/src/main/java/org/apache/hudi/io/HoodieCDCLogger.java

```diff
@@ -0,0 +1,263 @@
+import org.apache.hudi.common.table.cdc.HoodieCDCOperation;
+import org.apache.hudi.common.table.cdc.HoodieCDCSupplementalLoggingMode;
+import org.apache.hudi.common.table.cdc.HoodieCDCUtils;
+import org.apache.hudi.common.table.log.AppendResult;
+import org.apache.hudi.common.table.log.HoodieLogFormat;
+import org.apache.hudi.common.table.log.block.HoodieCDCDataBlock;
+import org.apache.hudi.common.table.log.block.HoodieLogBlock;
+import org.apache.hudi.common.util.DefaultSizeEstimator;
+import org.apache.hudi.common.util.Option;
+import org.apache.hudi.common.util.StringUtils;
+import org.apache.hudi.common.util.collection.ExternalSpillableMap;
+import org.apache.hudi.config.HoodieWriteConfig;
+import org.apache.hudi.exception.HoodieException;
+import org.apache.hudi.exception.HoodieIOException;
+import org.apache.hudi.exception.HoodieUpsertException;
+
+import java.io.Closeable;
+import java.io.IOException;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.stream.Collectors;
+
+/**
+ * This class encapsulates all the cdc-writing functions.
+ */
+public class HoodieCDCLogger implements Closeable {
+
+  private final String commitTime;
+
+  private final String keyField;
+
+  private final Schema dataSchema;
+
+  private final boolean populateMetaFields;
+
+  // writer for cdc data
+  private final HoodieLogFormat.Writer cdcWriter;
+
+  private final boolean cdcEnabled;
+
+  private final HoodieCDCSupplementalLoggingMode cdcSupplementalLoggingMode;
+
+  private final Schema cdcSchema;
+
+  // the cdc data
+  private final Map<String, HoodieAvroPayload> cdcData;
+
+  public HoodieCDCLogger(
+      String commitTime,
+      HoodieWriteConfig config,
+      HoodieTableConfig tableConfig,
+      Schema schema,
+      HoodieLogFormat.Writer cdcWriter,
+      long maxInMemorySizeInBytes) {
+    try {
+      this.commitTime = commitTime;
+      this.dataSchema = HoodieAvroUtils.removeMetadataFields(schema);
+      this.populateMetaFields = config.populateMetaFields();
+      this.keyField = populateMetaFields ? HoodieRecord.RECORD_KEY_METADATA_FIELD
+          : tableConfig.getRecordKeyFieldProp();
+      this.cdcWriter = cdcWriter;
+
+      this.cdcEnabled = config.getBooleanOrDefault(HoodieTableConfig.CDC_ENABLED);
+      this.cdcSupplementalLoggingMode = HoodieCDCSupplementalLoggingMode.parse(
+          config.getStringOrDefault(HoodieTableConfig.CDC_SUPPLEMENTAL_LOGGING_MODE));
+
+      if (cdcSupplementalLoggingMode.equals(HoodieCDCSupplementalLoggingMode.WITH_BEFORE_AFTER)) {
+        this.cdcSchema = HoodieCDCUtils.CDC_SCHEMA;
+      } else if (cdcSupplementalLoggingMode.equals(HoodieCDCSupplementalLoggingMode.WITH_BEFORE)) {
+        this.cdcSchema = HoodieCDCUtils.CDC_SCHEMA_OP_RECORDKEY_BEFORE;
+      } else {
+        this.cdcSchema = HoodieCDCUtils.CDC_SCHEMA_OP_AND_RECORDKEY;
+      }
+
+      this.cdcData = new ExternalSpillableMap<>(
```

Comment:
No need to tackle in this PR, but want to call it out that accumulating records in memory will certainly be problematic from memory footprint (as well as GC) perspective.

Note, that Spark does process either via iteration where it doesn't hold records (also limited to the micro-batch size if underlying format supports batch reads, otherwise it will hold just 1 record at a time) in memory for longer than the RDD execution chain requires. The only accumulation point is Parquet writer which is much more efficient though: a) relying on encoding as well as compression, b) storing binary/serialized date in memory

Reply:
thank you for pointing this out. i will think deeply about it.